### PR TITLE
Refactors Consensus::enterCommittee

### LINKED ISSUE #147: Refactor: Consensus::enterCommittee
- Update the function signature
```solidity
function enterCommittee(
    bytes32 _committeeMetachainId,
    bytes32 _validatorMetachainId,
    uint256 _validatorMetablockHeight,
    address _validator,
    address _furtherMember
)
    external
```

- The `bytes32 _committeeMetachainId` is used for the following
	- Determine for which metachain the Committee was formed.
	- To validate that the metablock round status was CommitteeFormed

- The `bytes32 _validatorMetachainId` is used for the following
	- Determine the metachain that validator belongs to.
	- To check when the validator joined the core. The validator can enter the committee if the validator has joined the core at a block height that was less than committee formation height.
	- To verify that the validator is still active.

- The `uint256 _validatorMetablockHeight` is required
  - This determines the latest metablockchain height for which the validator is active. The metachain for this validator can be different from the metachain for which the committee is formed.

- The `address _validator` is the address of the validator which has not been added and is required 
  - To check if the `_validator` is not slashed
  - To check if the `_validator` is valid  when the committee is formed

- The `address _furtherMember` is 
  - It is the nearest validator address in the committee.

- Reference: https://github.com/mosaicdao/mosaic-pm/blob/master/specification-discussions/20191206-consensus.md
  
  - For now, do not implement 
    - `address payable _reimbursementReceiver`, and
    - `reimburseGas(GAS_ENTER_COMMITTEE, _reimbursementReceiver)` at the function signature as these will be defined later

### PR DESCRIPTION
**Pull request by pgev**

### CODE REVIEW

**Comment by abhayks1** [comment]
@pgev Is it available for review?
Travis tests are red 🔴

**Comment by pgev** [comment]
@abhayks1 I've rebased and fixed failing tests. It's again ready for review.

File: contracts/core/Core.sol [hunk 1]
```diff
@@ -862,7 +859,25 @@ contract Core is MasterCopyNonUpgradable, ConsensusModule, MosaicVersion, CoreSt
...
+            !reputation.isSlashed(_account);
+    }
+
+    /**
+     * @notice Validator is active if the given metablock's height is:
+     *           - greater or equal than validator's begin height
+     *           - and, less or equal than validator's end height
+     *           - was not slashed
+     */
+    function isValidator(address _account, uint256 _metablockHeight)
```
Comment (abhayks1) [inline contracts/core/Core.sol:887]: A lot of code between `isActiveValidator` and `isValidator` is common. We can refactor the common code to a private method with parameters `account, metablockHeight`. isActiveValidator can pass` openKernelHeight` as metablockHeight

What do you think?

File: contracts/core/Core.sol [hunk 2]
```diff
@@ -862,7 +859,25 @@ contract Core is MasterCopyNonUpgradable, ConsensusModule, MosaicVersion, CoreSt
...
+            (validatorEndHeight[_account] > uint256(0)) &&
+            !reputation.isSlashed(_account);
+    }
+
+    /**
+     * @notice Validator is active if the given metablock's height is:
+     *           - greater or equal than validator's begin height
+     *           - and, less or equal than validator's end height
+     *           - was not slashed
+     */
```
Comment (abhayks1) [inline contracts/core/Core.sol:886]: We should add documentation for function arguments and and return values as per dev guidelines. 
Similar comment for isActiveValidator method.

File: contracts/consensus/Consensus.sol [hunk 3]
```diff
@@ -418,37 +418,58 @@ contract Consensus is MasterCopyNonUpgradable, CoreLifetimeEnum, MosaicVersion,
...
+     * @param _furtherMember Validator's address that is further member
+     *                       compared to the `_validator`.
      */
     function enterCommittee(
-        bytes32 _metachainId,
-        CoreI _core,
+        bytes32 _committeeMetachainId,
+        bytes32 _validatorMetachainId,
+        uint256 _validatorMetablockHeight,
         address _validator,
```
Comment (abhayks1) [inline contracts/consensus/Consensus.sol:548]: I feel `_validator` argument should come before `_validatorMetachainId`, _`validatorMetablockHeight`.
A validator is entering into committee. `_validatorMetachainId` and `_validatorMetablockHeight` seems like they are the properties of a validator. The `_validator` argument significance is strong enough to be prefixed before `_validatorMetachainId` and `_validatorMetablockHeight`
@pgev @deepesh-kn what do you think? It's an interface change.

File: contracts/consensus/Consensus.sol [hunk 4]
```diff
@@ -461,25 +482,66 @@ contract Consensus is MasterCopyNonUpgradable, CoreLifetimeEnum, MosaicVersion,
...
-            "Invalid validator was given."
+            validatorMetablock.round == MetablockRound.Committed,
+            "Provided metablock for validator must be committed."
         );
 
-        uint256 currentHeight = metablockTips[_metachainId];
-        Metablock storage currentMetablock = metablockchains[_metachainId][currentHeight];
+        uint256 formationHeight = committeeMetablock.roundBlockNumber;
+        uint256 checkValidatorForHeight = 0;
+        if (validatorMetablock.roundBlockNumber >= formationHeight) {
```
Comment (abhayks1) [inline contracts/consensus/Consensus.sol:590]: I think the logic of evaluating `checkValidatorForHeight` can be refactored to private/internal method 🤔.
enterCommittee method has grown significantly.

**Review by abhayks1 — CHANGES_REQUESTED** [review]
Nice work 👍 
Please see inline comments.

**Comment by pgev** [comment]
Closing the PR for now. Will add into #147 as a note for later reference if needed.

**Pull request closed without merging**